<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tidy Transcriptomics for pseudobulk Sequencing Analyses • tidyomicsWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tidy Transcriptomics for pseudobulk Sequencing Analyses">
<meta property="og:description" content="tidyomicsWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidyomicsWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.16.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pseudobulk_transcriptomics.html">Tidy Transcriptomics for pseudobulk Sequencing Analyses</a>
    </li>
    <li>
      <a href="../articles/single_cell_bioconductor_transcriptomics.html">Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Bioconductor</a>
    </li>
    <li>
      <a href="../articles/single_cell_seurat_transcriptomics.html">Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Seurat</a>
    </li>
    <li>
      <a href="../articles/solutions_transcriptomics.html">Solutions</a>
    </li>
    <li>
      <a href="../articles/supplementary_transcriptomics.html">Supplementary Material</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tidy-biology/TidyGenomicsTranscriptomicsWorkshop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tidy Transcriptomics for pseudobulk Sequencing
Analyses</h1>
                        <h4 data-toc-skip class="author">Maria Doyle,
Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 data-toc-skip class="author">Stefano
Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/tidy-biology/TidyGenomicsTranscriptomicsWorkshop/blob/HEAD/vignettes/pseudobulk_transcriptomics.Rmd" class="external-link"><code>vignettes/pseudobulk_transcriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>pseudobulk_transcriptomics.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="instructors">Instructors<a class="anchor" aria-label="anchor" href="#instructors"></a>
</h3>
<p><em>Dr. Stefano Mangiola</em> is currently a Postdoctoral researcher
in the laboratory of Prof. Tony Papenfuss at the Walter and Eliza Hall
Institute in Melbourne, Australia. His background spans from
biotechnology to bioinformatics and biostatistics. His research focuses
on prostate and breast tumour microenvironment, the development of
statistical models for the analysis of RNA sequencing data, and data
analysis and visualisation interfaces.</p>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 id="what-you-will-learn">What you will learn<a class="anchor" aria-label="anchor" href="#what-you-will-learn"></a>
</h4>
<ul>
<li>Basic <code>tidy</code> operations possible with
<code>tidyseurat</code> and <code>tidySingleCellExperiment</code>
</li>
<li>The differences between <code>Seurat</code> and
<code>SingleCellExperiment</code> representation, and <code>tidy</code>
representation</li>
<li>How to interface <code>Seurat</code> and
<code>SingleCellExperiment</code> with tidy manipulation and
visualisation</li>
<li>A real-world case study that will showcase the power of
<code>tidy</code> single-cell methods compared with base/ad-hoc
methods</li>
</ul>
</div>
<div class="section level4">
<h4 id="what-you-will-not-learn">What you will <em>not</em> learn<a class="anchor" aria-label="anchor" href="#what-you-will-not-learn"></a>
</h4>
<ul>
<li>The molecular technology of single-cell sequencing</li>
<li>The fundamentals of single-cell data analysis</li>
<li>The fundamentals of tidy data analysis</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h3>
<div class="section level4">
<h4 id="local">Local<a class="anchor" aria-label="anchor" href="#local"></a>
</h4>
<p>We will use the Cloud during the workshop and this method is
available if you want to run the material after the workshop. If you
want to install on your own computer, see instructions <a href="https://tidybiology.github.io/tidyomicsWorkshop/index.html#workshop-package-installation" class="external-link">here</a>.</p>
<p>Alternatively, you can view the material at the workshop webpage <a href="https://tidybiology.github.io/tidyomicsWorkshop/articles/main.html" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="introduction-to-tidytranscriptomics">Introduction to tidytranscriptomics<a class="anchor" aria-label="anchor" href="#introduction-to-tidytranscriptomics"></a>
</h3>
<p><a href="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidybiology/tidyomicsWorkshop/master/inst/tidytranscriptomics_slides.pdf" class="external-link">Here</a></p>
<iframe src="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidybiology/tidyomicsWorkshop/master/inst/tidytranscriptomics_slides.pdf&amp;embedded=true" scrolling="yes" style="width:100%; height:600px;" frameborder="0">
</iframe>
</div>
<div class="section level2">
<h2 id="pseudobulk-analyses">Pseudobulk analyses<a class="anchor" aria-label="anchor" href="#pseudobulk-analyses"></a>
</h2>
<p>Next we want to identify genes whose transcription is affected by
treatment in this dataset, comparing treated and untreated patients. We
can do this with pseudobulk analysis. We aggregate cell-wise transcript
abundance into pseudobulk samples and can then perform hypothesis
testing using the very well established bulk RNA sequencing tools. For
example, we can use DESeq2 in tidybulk to perform differential
expression testing. For more details on pseudobulk analysis see <a href="https://hbctraining.github.io/scRNA-seq/lessons/pseudobulk_DESeq2_scrnaseq.html" class="external-link">here</a>.</p>
<p>We want to do it for each cell type and the tidy transcriptomics
ecosystem makes this very easy.</p>
<div class="section level4">
<h4 id="create-pseudobulk-samples">Create pseudobulk samples<a class="anchor" aria-label="anchor" href="#create-pseudobulk-samples"></a>
</h4>
<p>To create pseudobulk samples from the single cell samples, we will
use a helper function called <code>aggregate_cells</code>, available in
this workshop package. This function will combine the single cells into
a group for each cell type for each sample.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu">tidyomicsWorkshop</span><span class="fu">::</span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">aggregate_cells</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sample</span>, <span class="va">cell_type</span><span class="op">)</span>, assays <span class="op">=</span> <span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>  <span class="fu">as_SummarizedExperiment</span><span class="op">(</span><span class="va">.sample</span>, <span class="va">.feature</span>, <span class="va">RNA</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="tidybulk-and-tidysummarizedexperiment">Tidybulk and tidySummarizedExperiment<a class="anchor" aria-label="anchor" href="#tidybulk-and-tidysummarizedexperiment"></a>
</h4>
<p>With <code>tidySummarizedExperiment</code> and <code>tidybulk</code>
it is easy to split the data into groups and perform analyses on each
without needing to create separate objects.</p>
</div>
<div class="section level3">
<h3 id="tidybulk-functionsutilities-available">Tidybulk functions/utilities available<a class="anchor" aria-label="anchor" href="#tidybulk-functionsutilities-available"></a>
</h3>
<table class="table">
<colgroup>
<col width="30%">
<col width="69%">
</colgroup>
<thead><tr class="header">
<th>Function</th>
<th>Description</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>aggregate_duplicates</code></td>
<td>Aggregate abundance and annotation of duplicated transcripts in a
robust way</td>
</tr>
<tr class="even">
<td>
<code>identify_abundant</code> <code>keep_abundant</code>
</td>
<td>identify or keep the abundant genes</td>
</tr>
<tr class="odd">
<td><code>keep_variable</code></td>
<td>Filter for top variable features</td>
</tr>
<tr class="even">
<td><code>scale_abundance</code></td>
<td>Scale (normalise) abundance for RNA sequencing depth</td>
</tr>
<tr class="odd">
<td><code>reduce_dimensions</code></td>
<td>Perform dimensionality reduction (PCA, MDS, tSNE, UMAP)</td>
</tr>
<tr class="even">
<td><code>cluster_elements</code></td>
<td>Labels elements with cluster identity (kmeans, SNN)</td>
</tr>
<tr class="odd">
<td><code>remove_redundancy</code></td>
<td>Filter out elements with highly correlated features</td>
</tr>
<tr class="even">
<td><code>adjust_abundance</code></td>
<td>Remove known unwanted variation (Combat)</td>
</tr>
<tr class="odd">
<td><code>test_differential_abundance</code></td>
<td>Differential transcript abundance testing (DESeq2, edgeR, voom)</td>
</tr>
<tr class="even">
<td><code>deconvolve_cellularity</code></td>
<td>Estimated tissue composition (Cibersort, llsr, epic, xCell,
mcp_counter, quantiseq</td>
</tr>
<tr class="odd">
<td><code>test_differential_cellularity</code></td>
<td>Differential cell-type abundance testing</td>
</tr>
<tr class="even">
<td><code>test_stratification_cellularity</code></td>
<td>Estimate Kaplan-Meier survival differences</td>
</tr>
<tr class="odd">
<td><code>test_gene_enrichment</code></td>
<td>Gene enrichment analyses (EGSEA)</td>
</tr>
<tr class="even">
<td><code>test_gene_overrepresentation</code></td>
<td>Gene enrichment on list of transcript names (no rank)</td>
</tr>
<tr class="odd">
<td><code>test_gene_rank</code></td>
<td>Gene enrichment on list of transcript (GSEA)</td>
</tr>
<tr class="even">
<td><code>impute_missing_abundance</code></td>
<td>Impute abundance for missing data points using sample groupings</td>
</tr>
</tbody>
</table>
<p>We use tidyverse <code>nest</code> to group the data. The command
below will create a tibble containing a column with a
SummarizedExperiment object for each cell type. <code>nest</code> is
similar to tidyverse <code>group_by</code>, except with
<code>nest</code> each group is stored in a single row, and can be a
complex object such as a plot or SummarizedExperiment.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span> <span class="op">&lt;-</span> </span>
<span>    <span class="va">pseudo_bulk</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">nest</span><span class="op">(</span>grouped_summarized_experiment <span class="op">=</span> <span class="op">-</span><span class="va">cell_type</span><span class="op">)</span></span>
<span></span>
<span><span class="va">pseudo_bulk_nested</span></span></code></pre></div>
<p>To explore the grouping, we can use tidyverse <code>slice</code> to
choose a row (cell_type) and <code>pull</code> to extract the values
from a column. If we pull the data column we can view the
SummarizedExperiment object.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">grouped_summarized_experiment</span><span class="op">)</span></span></code></pre></div>
<p>We can then identify differentially expressed genes for each cell
type for our condition of interest, treated versus untreated patients.
We use tidyverse <code>map</code> to apply differential expression
functions to each cell type group in the nested data. The result columns
will be added to the SummarizedExperiment objects.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Differential transcription abundance</span></span>
<span><span class="va">pseudo_bulk_nested</span> <span class="op">&lt;-</span></span>
<span>    </span>
<span>    <span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># map accepts a data column (.x) and a function. It applies the function to each element of the column.</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>grouped_summarized_experiment <span class="op">=</span> <span class="fu">map</span><span class="op">(</span></span>
<span>        <span class="va">grouped_summarized_experiment</span>,</span>
<span>        <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span></span>
<span>            </span>
<span>            <span class="co"># Removing genes with low expression</span></span>
<span>            <span class="fu">keep_abundant</span><span class="op">(</span>factor_of_interest <span class="op">=</span> <span class="va">treatment</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>            </span>
<span>            <span class="co"># Testing for differential expression using DESeq2  </span></span>
<span>            <span class="fu">test_differential_abundance</span><span class="op">(</span><span class="op">~</span><span class="va">treatment</span>, method<span class="op">=</span><span class="st">"DESeq2"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            </span>
<span>            <span class="co"># Scale abundance for FUTURE visualisation</span></span>
<span>            <span class="fu">scale_abundance</span><span class="op">(</span>method<span class="op">=</span><span class="st">"TMMwsp"</span><span class="op">)</span> </span>
<span>    <span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The output is again a tibble containing a SummarizedExperiment object
for each cell type.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span></span></code></pre></div>
<p>If we pull out the SummarizedExperiment object for the first cell
type, as before, we can see it now has columns containing the
differential expression results (e.g. logFC, PValue).</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">slice</span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">pull</span><span class="op">(</span><span class="va">grouped_summarized_experiment</span><span class="op">)</span></span></code></pre></div>
<p>We can analyse our nested dataset mapping queries across the
<code>SummarizedExperiment</code>s</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span> <span class="op">=</span> </span>
<span>    <span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Identify top significant genes</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>top_genes <span class="op">=</span> <span class="fu">map_chr</span><span class="op">(</span></span>
<span>        <span class="va">grouped_summarized_experiment</span>, </span>
<span>        <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu">pivot_transcript</span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu">arrange</span><span class="op">(</span><span class="va">pvalue</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>            <span class="fu">pull</span><span class="op">(</span><span class="va">.feature</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    </span>
<span>    <span class="co"># Filter top gene</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>grouped_summarized_experiment <span class="op">=</span> <span class="fu">map2</span><span class="op">(</span></span>
<span>        <span class="va">grouped_summarized_experiment</span>, <span class="va">top_genes</span>,</span>
<span>        <span class="op">~</span> <span class="fu"><a href="https://rdrr.io/r/stats/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">.x</span>, <span class="va">.feature</span> <span class="op">==</span> <span class="va">.y</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">pseudo_bulk_nested</span></span></code></pre></div>
<p>Plot top differential genes</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">pseudo_bulk_nested</span> <span class="op">=</span> </span>
<span>    <span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Plot significant genes for each cell type</span></span>
<span>    <span class="co"># map2 is map that accepts 2 input columns (.x, .y) and a function</span></span>
<span>    <span class="fu">mutate</span><span class="op">(</span>plot <span class="op">=</span> <span class="fu">map2</span><span class="op">(</span></span>
<span>        <span class="va">grouped_summarized_experiment</span>,<span class="va">cell_type</span>,</span>
<span>        <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span></span>
<span>            </span>
<span>            <span class="co"># Plot</span></span>
<span>            <span class="fu">ggplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span><span class="va">treatment</span>, <span class="va">RNA_scaled</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">geom_boxplot</span><span class="op">(</span><span class="fu">aes</span><span class="op">(</span>fill <span class="op">=</span> <span class="va">treatment</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">geom_jitter</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">scale_y_log10</span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">facet_wrap</span><span class="op">(</span><span class="op">~</span><span class="va">.feature</span>, ncol <span class="op">=</span> <span class="fl">3</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">ggtitle</span><span class="op">(</span><span class="va">.y</span><span class="op">)</span> <span class="op">+</span></span>
<span>            <span class="fu">tidyomicsWorkshop</span><span class="fu">::</span><span class="va"><a href="../reference/theme_multipanel.html">theme_multipanel</a></span></span>
<span>    <span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">pseudo_bulk_nested</span></span>
<span></span>
<span><span class="va">pseudo_bulk_nested</span> <span class="op">|&gt;</span> <span class="fu">pull</span><span class="op">(</span><span class="va">plot</span><span class="op">)</span> </span></code></pre></div>
<p><strong>Session Information</strong></p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>References</strong></p>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
