<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Seurat • tidyomicsWorkshop</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/flatly/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Seurat">
<meta property="og:description" content="tidyomicsWorkshop">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">tidyomicsWorkshop</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.16.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/pseudobulk_transcriptomics.html">Tidy Transcriptomics for pseudobulk Sequencing Analyses</a>
    </li>
    <li>
      <a href="../articles/single_cell_bioconductor_transcriptomics.html">Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Bioconductor</a>
    </li>
    <li>
      <a href="../articles/single_cell_seurat_transcriptomics.html">Tidy Transcriptomics for Single-cell RNA Sequencing Analyses with Seurat</a>
    </li>
    <li>
      <a href="../articles/solutions_transcriptomics.html">Solutions</a>
    </li>
    <li>
      <a href="../articles/supplementary_transcriptomics.html">Supplementary Material</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/tidy-biology/TidyGenomicsTranscriptomicsWorkshop" class="external-link">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Tidy Transcriptomics for Single-cell RNA
Sequencing Analyses with Seurat</h1>
                        <h4 data-toc-skip class="author">Maria Doyle,
Peter MacCallum Cancer Centre<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>
</h4>
                        <h4 data-toc-skip class="author">Stefano
Mangiola, Walter and Eliza Hall Institute<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a>
</h4>
            
      
      <small class="dont-index">Source: <a href="https://github.com/tidy-biology/TidyGenomicsTranscriptomicsWorkshop/blob/HEAD/vignettes/single_cell_seurat_transcriptomics.Rmd" class="external-link"><code>vignettes/single_cell_seurat_transcriptomics.Rmd</code></a></small>
      <div class="hidden name"><code>single_cell_seurat_transcriptomics.Rmd</code></div>

    </div>

    
    
<div class="section level3">
<h3 id="instructors">Instructors<a class="anchor" aria-label="anchor" href="#instructors"></a>
</h3>
<p><em>Dr. Stefano Mangiola</em> is currently a Postdoctoral researcher
in the laboratory of Prof. Tony Papenfuss at the Walter and Eliza Hall
Institute in Melbourne, Australia. His background spans from
biotechnology to bioinformatics and biostatistics. His research focuses
on prostate and breast tumour microenvironment, the development of
statistical models for the analysis of RNA sequencing data, and data
analysis and visualisation interfaces.</p>
</div>
<div class="section level3">
<h3 id="workshop-goals-and-objectives">Workshop goals and objectives<a class="anchor" aria-label="anchor" href="#workshop-goals-and-objectives"></a>
</h3>
<div class="section level4">
<h4 id="what-you-will-learn">What you will learn<a class="anchor" aria-label="anchor" href="#what-you-will-learn"></a>
</h4>
<ul>
<li>Basic <code>tidy</code> operations possible with
<code>tidyseurat</code> and <code>tidySingleCellExperiment</code>
</li>
<li>The differences between <code>Seurat</code> and
<code>SingleCellExperiment</code> representation, and <code>tidy</code>
representation</li>
<li>How to interface <code>Seurat</code> and
<code>SingleCellExperiment</code> with tidy manipulation and
visualisation</li>
<li>A real-world case study that will showcase the power of
<code>tidy</code> single-cell methods compared with base/ad-hoc
methods</li>
</ul>
</div>
<div class="section level4">
<h4 id="what-you-will-not-learn">What you will <em>not</em> learn<a class="anchor" aria-label="anchor" href="#what-you-will-not-learn"></a>
</h4>
<ul>
<li>The molecular technology of single-cell sequencing</li>
<li>The fundamentals of single-cell data analysis</li>
<li>The fundamentals of tidy data analysis</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="getting-started">Getting started<a class="anchor" aria-label="anchor" href="#getting-started"></a>
</h3>
<div class="section level4">
<h4 id="local">Local<a class="anchor" aria-label="anchor" href="#local"></a>
</h4>
<p>We will use the Cloud during the workshop and this method is
available if you want to run the material after the workshop. If you
want to install on your own computer, see instructions <a href="https://tidybiology.github.io/tidyomicsWorkshop/index.html#workshop-package-installation" class="external-link">here</a>.</p>
<p>Alternatively, you can view the material at the workshop webpage <a href="https://tidybiology.github.io/tidyomicsWorkshop/articles/main.html" class="external-link">here</a>.</p>
</div>
</div>
<div class="section level3">
<h3 id="introduction-to-tidytranscriptomics">Introduction to tidytranscriptomics<a class="anchor" aria-label="anchor" href="#introduction-to-tidytranscriptomics"></a>
</h3>
<p><a href="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidybiology/tidyomicsWorkshop/master/inst/tidytranscriptomics_slides.pdf" class="external-link">Here</a></p>
<iframe src="https://docs.google.com/gview?url=https://raw.githubusercontent.com/tidybiology/tidyomicsWorkshop/master/inst/tidytranscriptomics_slides.pdf&amp;embedded=true" scrolling="yes" style="width:100%; height:600px;" frameborder="0">
</iframe>
</div>
<div class="section level2">
<h2 id="part-1-introduction-to-tidyseurat">Part 1 Introduction to tidyseurat<a class="anchor" aria-label="anchor" href="#part-1-introduction-to-tidyseurat"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Load packages</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://purrr.tidyverse.org/" class="external-link">purrr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://dplyr.tidyverse.org" class="external-link">dplyr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://colorspace.R-Forge.R-project.org/" class="external-link">colorspace</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">dittoSeq</span><span class="op">)</span></span></code></pre></div>
<p>Seurat is a very popular analysis toolkit for single cell RNA
sequencing data <span class="citation">(Butler et al. 2018; Stuart et
al. 2019)</span>.</p>
<p>Here we load single-cell data in Seurat object format. This data is
peripheral blood mononuclear cells (PBMCs) from metastatic breast cancer
patients.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># load single cell RNA sequencing data</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="fu">tidyomicsWorkshop</span><span class="fu">::</span><span class="va">seurat_obj</span></span>
<span></span>
<span><span class="co"># take a look</span></span>
<span><span class="va">seurat_obj</span></span></code></pre></div>
<p>tidyseurat provides a bridge between the Seurat single-cell package
and the tidyverse <span class="citation">(Wickham et al. 2019)</span>.
It creates an invisible layer that enables viewing the Seurat object as
a tidyverse tibble, and provides Seurat-compatible <em>dplyr</em>,
<em>tidyr</em>, <em>ggplot</em> and <em>plotly</em> functions.</p>
<p>If we load the <em>tidyseurat</em> package and then view the single
cell data, it now displays as a tibble.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/stemangiola/tidyseurat" class="external-link">tidyseurat</a></span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj</span> </span></code></pre></div>
<p>If we want to revert to the standard SingleCellExperiment view we can
do that.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_Seurat_show"</span> <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span></span></code></pre></div>
<p>If we want to revert back to tidy SingleCellExperiment view we
can.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"restore_Seurat_show"</span> <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="va">seurat_obj</span></span></code></pre></div>
<p>It can be interacted with using <a href="https://satijalab.org/seurat/articles/essential_commands.html" class="external-link">Seurat
commands</a> such as <code>Assays</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/Assays-class.html" class="external-link">Assays</a></span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span></span></code></pre></div>
<p>We can also interact with our object as we do with any tidyverse
tibble.</p>
<div class="section level4">
<h4 id="tidyverse-commands">Tidyverse commands<a class="anchor" aria-label="anchor" href="#tidyverse-commands"></a>
</h4>
<p>We can use tidyverse commands, such as <code>filter</code>,
<code>select</code> and <code>mutate</code> to explore the tidyseurat
object. Some examples are shown below and more can be seen at the
tidyseurat website <a href="https://stemangiola.github.io/tidyseurat/articles/introduction.html#tidyverse-commands-1" class="external-link">here</a>.</p>
<p>We can use <code>filter</code> to choose rows, for example, to see
just the rows for the cells in G1 cell-cycle stage. Check if have groups
or ident present.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">Phase</span> <span class="op">==</span> <span class="st">"G1"</span><span class="op">)</span></span></code></pre></div>
<p>We can use <code>select</code> to choose columns, for example, to see
the sample, cell, total cellular RNA</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">nCount_RNA</span>, <span class="va">Phase</span><span class="op">)</span></span></code></pre></div>
<p>We also see the UMAP columns as they are not part of the cell
metadata, they are read-only. If we want to save the edited metadata,
the Seurat object is modified accordingly.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Save edited metadata</span></span>
<span><span class="va">seurat_modified</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">nCount_RNA</span>, <span class="va">Phase</span><span class="op">)</span></span>
<span><span class="co"># View Seurat metadata</span></span>
<span><span class="va">seurat_modified</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p>We can use <code>mutate</code> to create a column. For example, we
could create a new <code>Phase_l</code> column that contains a
lower-case version of <code>Phase</code>.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>Phase_l<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/chartr.html" class="external-link">tolower</a></span><span class="op">(</span><span class="va">Phase</span><span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Select columns to view    </span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">Phase</span>, <span class="va">Phase_l</span><span class="op">)</span></span></code></pre></div>
<p>We can use tidyverse commands to polish an annotation column. We will
extract the sample, and group information from the file name column into
separate columns.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># First take a look at the file column</span></span>
<span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">file</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Create columns for sample and group</span></span>
<span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Extract sample and group</span></span>
<span>    <span class="fu"><a href="https://tidyr.tidyverse.org/reference/extract.html" class="external-link">extract</a></span><span class="op">(</span><span class="va">file</span>, <span class="st">"sample"</span>, <span class="st">"../data/.*/([a-zA-Z0-9_-]+)/outs.+"</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look</span></span>
<span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">sample</span><span class="op">)</span></span></code></pre></div>
<p>We could use tidyverse <code>unite</code> to combine columns, for
example to create a new column for sample id that combines the sample
and patient identifier (BCB) columns.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://tidyr.tidyverse.org/reference/unite.html" class="external-link">unite</a></span><span class="op">(</span><span class="st">"sample_id"</span>, <span class="va">sample</span>, <span class="va">BCB</span>, remove <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Take a look</span></span>
<span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">.cell</span>, <span class="va">sample_id</span>, <span class="va">sample</span>, <span class="va">BCB</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
<div class="section level2">
<h2 id="part-2-signature-visualisation">Part 2 Signature visualisation<a class="anchor" aria-label="anchor" href="#part-2-signature-visualisation"></a>
</h2>
<p>We will now demonstrate a real-world example of the power of using
tidy transcriptomics packages in single cell analysis. For more
information on single-cell analysis steps performed in a tidy way please
see the <a href="https://tidytranscriptomics-workshops.github.io/ismb2021_tidytranscriptomics/articles/tidytranscriptomics.html" class="external-link">ISMB2021
workshop</a>.</p>
<div class="section level4">
<h4 id="data-pre-processing">Data pre-processing<a class="anchor" aria-label="anchor" href="#data-pre-processing"></a>
</h4>
<p>The object <code>seurat_obj</code> we’ve been using was created as
part of a study on breast cancer systemic immune response. Peripheral
blood mononuclear cells have been sequenced for RNA at the single-cell
level. The steps used to generate the object are summarised below.</p>
<ul>
<li><p><code>scran</code>, <code>scater</code>, and
<code>DropletsUtils</code> packages have been used to eliminate empty
droplets and dead cells. Samples were individually quality checked and
cells were filtered for good gene coverage.</p></li>
<li><p>Variable features were identified using
<code>Seurat</code>.</p></li>
<li><p>Read counts were scaled and normalised using SCTtransform from
<code>Seurat</code>.</p></li>
<li><p>Data integration was performed using <code>Seurat</code> with
default parameters.</p></li>
<li><p>PCA performed to reduce feature dimensionality.</p></li>
<li><p>Nearest-neighbor cell networks were calculated using 30 principal
components.</p></li>
<li><p>2 UMAP dimensions were calculated using 30 principal
components.</p></li>
<li><p>Cells with similar transcriptome profiles were grouped into
clusters using Louvain clustering from <code>Seurat</code>.</p></li>
</ul>
</div>
<div class="section level4">
<h4 id="analyse-custom-signature">Analyse custom signature<a class="anchor" aria-label="anchor" href="#analyse-custom-signature"></a>
</h4>
<p>The researcher analysing this dataset wanted to to identify gamma
delta T cells using a gene signature from a published paper <span class="citation">(Pizzolato et al. 2019)</span>.</p>
<p>With tidyseurat’s <code>join_features</code> the counts for the genes
could be viewed as columns.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>    <span class="op">)</span> </span></code></pre></div>
<div class="section level5">
<h5 id="signature-calculation">Signature calculation<a class="anchor" aria-label="anchor" href="#signature-calculation"></a>
</h5>
<p>They were able to use tidyseurat’s <code>join_features</code> to
select the counts for the genes in the signature, followed by tidyverse
<code>mutate</code> to easily create a column containing the signature
score.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>        </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signature_score <span class="op">=</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span> <span class="op">+</span> <span class="va">TRGC2</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="va">signature_score</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>The gamma delta T cells could then be visualised by the signature
score using Seurat’s visualisation functions.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>        </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signature_score <span class="op">=</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span><span class="op">+</span> <span class="va">TRGC2</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu">Seurat</span><span class="fu">::</span><span class="fu">FeaturePlot</span><span class="op">(</span>features <span class="op">=</span>  <span class="st">"signature_score"</span>, min.cutoff <span class="op">=</span> <span class="fl">0</span><span class="op">)</span> </span></code></pre></div>
<p>The cells could also be visualised using the popular and powerful
ggplot package, enabling the researcher to use ggplot functions they
were familiar with, and to customise the plot with great
flexibility.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>        </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signature_score <span class="op">=</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span><span class="op">+</span> <span class="va">TRGC2</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># plot cells with high score last so they're not obscured by other cells</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">signature_score</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span><span class="va">UMAP_1</span>, <span class="va">UMAP_2</span>, color <span class="op">=</span> <span class="va">signature_score</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size<span class="op">=</span><span class="fl">0.5</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_brewer.html" class="external-link">scale_color_distiller</a></span><span class="op">(</span>palette <span class="op">=</span> <span class="st">"Spectral"</span><span class="op">)</span> <span class="op">+</span></span>
<span>    <span class="fu">tidyomicsWorkshop</span><span class="fu">::</span><span class="va"><a href="../reference/theme_multipanel.html">theme_multipanel</a></span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="filtering-based-on-gate">Filtering based on gate<a class="anchor" aria-label="anchor" href="#filtering-based-on-gate"></a>
</h5>
<p>The gamma delta T cells (the blue cluster on the left with high
signature score) could be interactively selected from the plot using the
tidygate package.</p>
<p>This code can only be executed interactively from an R file and not
from a markdown file. So, in this casre, we need to copy and paste this
into the console.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_gamma_delta</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>        </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signature_score <span class="op">=</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span><span class="op">+</span> <span class="va">TRGC2</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>gate <span class="op">=</span> <span class="fu">tidygate</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/tidygate/man/gate_chr-methods.html" class="external-link">gate_int</a></span><span class="op">(</span></span>
<span>        <span class="va">UMAP_1</span>, <span class="va">UMAP_2</span>, </span>
<span>        .size <span class="op">=</span> <span class="fl">0.1</span>, </span>
<span>        .color <span class="op">=</span><span class="va">signature_score</span></span>
<span>    <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">gate</span><span class="op">==</span><span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="filtering-based-on-threshold">Filtering based on threshold<a class="anchor" aria-label="anchor" href="#filtering-based-on-threshold"></a>
</h5>
<p>For exploratory analyses, we can select the gamma delta T cells, the
red cluster on the left with high signature score. We’ll filter for
cells with a signature score &gt; 0.7.</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_gamma_delta</span> <span class="op">&lt;-</span></span>
<span>    </span>
<span>    <span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>, shape <span class="op">=</span> <span class="st">"wide"</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span></span>
<span>        signature_score <span class="op">=</span></span>
<span>            <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span> <span class="op">+</span> <span class="va">TRGC2</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>            <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Proper cluster selection should be used instead (see supplementary material)</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signature_score</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level5">
<h5 id="reanalysis-after-filtering">Reanalysis after filtering<a class="anchor" aria-label="anchor" href="#reanalysis-after-filtering"></a>
</h5>
<p>It was then possible to perform analyses on these gamma delta T cells
by simply chaining further commands, such as below.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/source.html" class="external-link">source</a></span><span class="op">(</span><span class="st">"https://raw.githubusercontent.com/satijalab/seurat-wrappers/master/R/fast_mnn.R"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    </span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/ttservice/man/join_features.html" class="external-link">join_features</a></span><span class="op">(</span></span>
<span>        features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span>, <span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>,</span>
<span>        shape <span class="op">=</span> <span class="st">"wide"</span>,</span>
<span>        assay <span class="op">=</span> <span class="st">"SCT"</span></span>
<span>        </span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>signature_score <span class="op">=</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD3D</span> <span class="op">+</span> <span class="va">TRDC</span> <span class="op">+</span> <span class="va">TRGC1</span><span class="op">+</span> <span class="va">TRGC2</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span> <span class="op">-</span></span>
<span>                    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span><span class="va">CD8A</span> <span class="op">+</span> <span class="va">CD8B</span>, to<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>,<span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="co"># Proper cluster selection should be used instead (see supplementary material)</span></span>
<span>    <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">signature_score</span> <span class="op">&gt;</span> <span class="fl">0.5</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    </span>
<span>    <span class="co"># Reanalysis</span></span>
<span>    <span class="fu">NormalizeData</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">FindVariableFeatures</span><span class="op">(</span>nfeatures <span class="op">=</span> <span class="fl">100</span>, assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">SplitObject</span><span class="op">(</span>split.by <span class="op">=</span> <span class="st">"file"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">SeuratWrappers</span><span class="fu">::</span><span class="fu">RunFastMNN</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span>, features <span class="op">=</span> <span class="fl">100</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">RunUMAP</span><span class="op">(</span>reduction <span class="op">=</span> <span class="st">"mnn"</span>, dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">FindNeighbors</span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">20</span>, reduction <span class="op">=</span> <span class="st">"mnn"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">FindClusters</span><span class="op">(</span>resolution <span class="op">=</span> <span class="fl">0.3</span><span class="op">)</span></span></code></pre></div>
<p>For comparison, we show the alternative using base R and
SingleCellExperiment. Note that the code contains more redundancy and
intermediate objects.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">counts_positive</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD3D"</span>, <span class="st">"TRDC"</span>, <span class="st">"TRGC1"</span>, <span class="st">"TRGC2"</span><span class="op">)</span>, <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span>to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">counts_negative</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/pkg/SummarizedExperiment/man/SummarizedExperiment-class.html" class="external-link">assay</a></span><span class="op">(</span><span class="va">seurat_obj</span>, <span class="st">"logcounts"</span><span class="op">)</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"CD8A"</span>, <span class="st">"CD8B"</span><span class="op">)</span>, <span class="op">]</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/colSums.html" class="external-link">colSums</a></span><span class="op">(</span><span class="op">)</span> <span class="op">|&gt;</span></span>
<span>    <span class="fu">scales</span><span class="fu">::</span><span class="fu"><a href="https://scales.r-lib.org/reference/rescale.html" class="external-link">rescale</a></span><span class="op">(</span>to <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj</span><span class="op">$</span><span class="va">signature_score</span> <span class="op">&lt;-</span> <span class="va">counts_positive</span> <span class="op">-</span> <span class="va">counts_negative</span></span>
<span></span>
<span><span class="va">seurat_obj_gamma_delta</span> <span class="op">&lt;-</span> <span class="va">seurat_obj</span><span class="op">[</span>, <span class="va">seurat_obj</span><span class="op">$</span><span class="va">signature_score</span> <span class="op">&gt;</span> <span class="fl">0.7</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># Reanalysis</span></span>
<span><span class="co"># ...</span></span></code></pre></div>
</div>
</div>
<div class="section level4">
<h4 id="interactive-plotting">Interactive plotting<a class="anchor" aria-label="anchor" href="#interactive-plotting"></a>
</h4>
<p>As a final note, it’s also possible to do complex and powerful things
in a simple way, due to the integration of the tidy transcriptomics
packages with the tidy universe. As one example, we can visualise the
cells as a 3D plot using plotly.</p>
<p>The example data we’ve been using only contains a few genes, for the
sake of time and size in this demonstration, but below is how you could
generate the 3 dimensions needed for 3D plot with a full dataset.</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">single_cell_object</span> <span class="op">|&gt;</span> </span>
<span>    <span class="fu">RunUMAP</span><span class="op">(</span>dims <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">30</span>, n.components <span class="op">=</span> <span class="fl">3L</span>, spread <span class="op">=</span> <span class="fl">0.5</span>,min.dist  <span class="op">=</span> <span class="fl">0.01</span>, n.neighbors <span class="op">=</span> <span class="fl">10L</span><span class="op">)</span></span></code></pre></div>
<p>We’ll demonstrate creating a 3D plot using some data that has 3 UMAP
dimensions.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_UMAP3</span> <span class="op">&lt;-</span> <span class="fu">tidyomicsWorkshop</span><span class="fu">::</span><span class="va">seurat_obj_UMAP3</span></span>
<span></span>
<span><span class="co"># Look at the object</span></span>
<span><span class="va">seurat_obj_UMAP3</span> <span class="op">|&gt;</span> <span class="fu"><a href="https://dplyr.tidyverse.org/reference/select.html" class="external-link">select</a></span><span class="op">(</span><span class="fu"><a href="https://tidyselect.r-lib.org/reference/starts_with.html" class="external-link">contains</a></span><span class="op">(</span><span class="st">"UMAP"</span><span class="op">)</span>, <span class="fu"><a href="https://tidyselect.r-lib.org/reference/everything.html" class="external-link">everything</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj_UMAP3</span> <span class="op">|&gt;</span></span>
<span>    </span>
<span>    <span class="fu">tidyseurat</span><span class="fu">::</span><span class="fu">plot_ly</span><span class="op">(</span></span>
<span>        x <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_1`</span>,</span>
<span>        y <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_2`</span>,</span>
<span>        z <span class="op">=</span> <span class="op">~</span><span class="va">`UMAP_3`</span>,</span>
<span>        color <span class="op">=</span> <span class="op">~</span><span class="va">cell_type</span>,</span>
<span>        colors <span class="op">=</span> <span class="fu">dittoSeq</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/dittoSeq/man/dittoColors.html" class="external-link">dittoColors</a></span><span class="op">(</span><span class="op">)</span></span>
<span>    <span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>    </span>
<span>    <span class="fu">plotly</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/plotly/man/add_trace.html" class="external-link">add_markers</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/AsIs.html" class="external-link">I</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level4">
<h4 id="exercises">Exercises<a class="anchor" aria-label="anchor" href="#exercises"></a>
</h4>
<ol style="list-style-type: decimal">
<li><p>What proportion of all cells are gamma-delta T cells?</p></li>
<li><p>There is a cluster of cells characterised by a low RNA output
(nCount_RNA). Use tidygate to identify the cell composition (cell_type)
of that cluster.</p></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="part-3-nested-analyses">Part 3 Nested analyses<a class="anchor" aria-label="anchor" href="#part-3-nested-analyses"></a>
</h2>
<p>When analysing single cell data is sometimes necessary to perform
calculations on data subsets. For example, we might want to estimate
difference in mRNA abundance between two condition for each cell
type.</p>
<p><code>tidyr</code> and <code>purrr</code> offer a great tool to
perform iterativre analyses in a functional way.</p>
<p>We use tidyverse <code>nest</code> to group the data. The command
below will create a tibble containing a column with a
SummarizedExperiment object for each cell type. <code>nest</code> is
similar to tidyverse <code>group_by</code>, except with
<code>nest</code> each group is stored in a single row, and can be a
complex object such as <code>Seurat</code>.</p>
<p>First let’s have a look to the cell types that constitute this
dataset</p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/count.html" class="external-link">count</a></span><span class="op">(</span><span class="va">cell_type</span><span class="op">)</span></span></code></pre></div>
<p>Let’s group the cells based on cell identity using
<code>nest</code></p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Set idents for the differential analysis</span></span>
<span><span class="fu">Idents</span><span class="op">(</span><span class="va">seurat_obj</span><span class="op">)</span> <span class="op">=</span> <span class="va">seurat_obj</span><span class="op">[[</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="st">"treatment"</span></span>
<span></span>
<span><span class="va">seurat_obj_nested</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>seurat <span class="op">=</span> <span class="op">-</span><span class="va">cell_type</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">seurat_obj_nested</span></span></code></pre></div>
<p>Let’s see what the first element of the Surat column looks like</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_nested</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">seurat</span><span class="op">)</span></span></code></pre></div>
<p>Now, let’s perform a differential gene-transcript abundance analysis
between the two conditions for each cell type.</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_nested</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj_nested</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Select significant genes</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>significant_genes <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat</span>,</span>
<span>    <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> </span>
<span>      </span>
<span>      <span class="co"># Test</span></span>
<span>      <span class="fu">NormalizeData</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu">FindAllMarkers</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      </span>
<span>      <span class="co"># Select top genes</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p_val_adj</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="va">seurat_obj_nested</span></span></code></pre></div>
<p>We can the lies the top genes with the heat map iteratively across
the cell types</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_nested</span> <span class="op">=</span> </span>
<span>  <span class="va">seurat_obj_nested</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Build heatmaps</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>heatmap <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html" class="external-link">map2</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat</span>, <span class="va">significant_genes</span>,</span>
<span>    <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> </span>
<span>       <span class="fu">ScaleData</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu">DoHeatmap</span><span class="op">(</span><span class="va">.y</span>, assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> </span>
<span></span>
<span><span class="va">seurat_obj_nested</span></span></code></pre></div>
<p>Let’s have a look to the first heatmap</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj_nested</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/slice.html" class="external-link">slice</a></span><span class="op">(</span><span class="fl">1</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">heatmap</span><span class="op">)</span></span></code></pre></div>
<p>You can do this whole analysis without saving any temporary variable
using the piping functionality of tidy R programming</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">seurat_obj</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Nest</span></span>
<span>  <span class="fu"><a href="https://tidyr.tidyverse.org/reference/nest.html" class="external-link">nest</a></span><span class="op">(</span>seurat <span class="op">=</span> <span class="op">-</span><span class="va">cell_type</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Select significant genes</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>significant_genes <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map.html" class="external-link">map</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat</span>,</span>
<span>    <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> </span>
<span>      </span>
<span>      <span class="co"># Test</span></span>
<span>      <span class="fu">NormalizeData</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu">FindAllMarkers</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      </span>
<span>      <span class="co"># Select top genes</span></span>
<span>      <span class="fu"><a href="https://dplyr.tidyverse.org/reference/filter.html" class="external-link">filter</a></span><span class="op">(</span><span class="va">p_val_adj</span><span class="op">&lt;</span><span class="fl">0.05</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>      <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Build heatmaps</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/mutate.html" class="external-link">mutate</a></span><span class="op">(</span>heatmap <span class="op">=</span> <span class="fu"><a href="https://purrr.tidyverse.org/reference/map2.html" class="external-link">map2</a></span><span class="op">(</span></span>
<span>    <span class="va">seurat</span>, <span class="va">significant_genes</span>,</span>
<span>    <span class="op">~</span> <span class="va">.x</span> <span class="op">|&gt;</span> </span>
<span>       <span class="fu">ScaleData</span><span class="op">(</span>assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>        <span class="fu">DoHeatmap</span><span class="op">(</span><span class="va">.y</span>, assay<span class="op">=</span><span class="st">"RNA"</span><span class="op">)</span></span>
<span>  <span class="op">)</span><span class="op">)</span> <span class="op">|&gt;</span> </span>
<span>  </span>
<span>  <span class="co"># Extract heatmaps</span></span>
<span>  <span class="fu"><a href="https://dplyr.tidyverse.org/reference/pull.html" class="external-link">pull</a></span><span class="op">(</span><span class="va">heatmap</span><span class="op">)</span></span></code></pre></div>
<div class="section level4">
<h4 id="exercises-1">Exercises<a class="anchor" aria-label="anchor" href="#exercises-1"></a>
</h4>
<ol style="list-style-type: decimal">
<li>Let’s suppose that you want to perform the analyses only for cell
types that have a total number of cells bigger than 1000. For example,
if a cell type has less than a sum of 1000 cells across all samples,
that cell type will be dropped from the dataset.</li>
</ol>
<ul>
<li>Answer this question avoiding to save temporary variables, and using
the function add_count to count the cells (before nesting), and then
filter</li>
<li>Answer this question avoiding to save temporary variables, and using
the function map_int to count the cells (after nesting), and the
filter</li>
</ul>
<p><strong>Session Information</strong></p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<p><strong>References</strong></p>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-butler2018integrating" class="csl-entry">
Butler, Andrew, Paul Hoffman, Peter Smibert, Efthymia Papalexi, and
Rahul Satija. 2018. <span>“Integrating Single-Cell Transcriptomic Data
Across Different Conditions, Technologies, and Species.”</span>
<em>Nature Biotechnology</em> 36 (5): 411–20.
</div>
<div id="ref-Pizzolato2019" class="csl-entry">
Pizzolato, Gabriele, Hannah Kaminski, Marie Tosolini, Don-Marc
Franchini, Fréderic Pont, Fréderic Martins, Carine Valle, et al. 2019.
<span>“Single-Cell <span>RNA</span> Sequencing Unveils the Shared and
the Distinct Cytotoxic Hallmarks of Human <span>TCRV</span><span class="math inline">\(\updelta\)</span>1 and <span>TCRV</span><span class="math inline">\(\updelta\)</span>2 <span class="math inline">\(\upgamma\)</span><span class="math inline">\(\updelta\)</span> t Lymphocytes.”</span>
<em>Proceedings of the National Academy of Sciences</em>, May,
201818488. <a href="https://doi.org/10.1073/pnas.1818488116" class="external-link">https://doi.org/10.1073/pnas.1818488116</a>.
</div>
<div id="ref-stuart2019comprehensive" class="csl-entry">
Stuart, Tim, Andrew Butler, Paul Hoffman, Christoph Hafemeister,
Efthymia Papalexi, William M Mauck III, Yuhan Hao, Marlon Stoeckius,
Peter Smibert, and Rahul Satija. 2019. <span>“Comprehensive Integration
of Single-Cell Data.”</span> <em>Cell</em> 177 (7): 1888–1902.
</div>
<div id="ref-wickham2019welcome" class="csl-entry">
Wickham, Hadley, Mara Averick, Jennifer Bryan, Winston Chang, Lucy
D’Agostino McGowan, Romain François, Garrett Grolemund, et al. 2019.
<span>“Welcome to the Tidyverse.”</span> <em>Journal of Open Source
Software</em> 4 (43): 1686.
</div>
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>&lt;maria.doyle at petermac.org&gt;<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>&lt;mangiola.s at wehi.edu.au&gt;<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Stefano Mangiola, Michael Love.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
